pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/IndexLoadingConfig.java

```diff
@@ -215,6 +218,26 @@ private void extractFromTableConfig(TableConfig tableConfig) {
     }
   }
 
+  /**
+   * Extracts compressionType for each column. Populates a map containing column name as key and compression type as
+   * value. Note that only RAW forward index columns will be populated in this map.
+   * @param tableConfig table config
+   */
+  private void extractCompressionConfigs(TableConfig tableConfig) {
```

Comment:
I think we need to handle the existing way compression is configured. There could be cases where people are configuring compressionCodec without going via FieldConfig (since this was added later and no migration hasn't happened).

Reply:
+1

Reply:
Existing CompresstionType is determined from the forwardIndex header. 
The New (Desired) CompresionType is fetched from fieldConfig.

**Reasoning for NOT using noDictionaryConfig to determine desired compressionType:**
I understand that we used noDictionaryConfig earlier to populate compression settings. The reasoning for relying only on the new fieldConfig approach is that - any user intending to use the "segmentReload with compressionChange" feature should go down the new way of setting compressionType in fieldConfigs instead of the old way. If we continue to add new feature support for changing compressionType with noDictionaryConfig, we would just be adding more tech debt that we would have to clean up later. Also, how would we resolve conflicts if user specifies two different values in noDictionaryConfig and fieldConfig?


IMO, this doesn't go against backward compatibility because all we are saying is we will support new features with only with fieldConfig. But, I understand the concern and would be happy to discuss this further to reach a conclusion.

Reply:
I see your point, but IMO we should first tackle migrating everyone and then cleaning up the code to stop looking at the older compression codec flag. We can perhaps establish a precedence ordering that if `fieldConfig` type compression codec is present it overrides the old config. The old mechanism is honored when adding a new column and for new segments that get created. So it's just a bit confusing that we honor it there but not here.

That said, I don't have a very strong opinion on this, since even if customers do the wrong thing by updating the older flag we can recommend them to use the new `fieldConfig` way to update the compression codec.

Reply:
My concern was the following

- User set the `noDictionaryConfig` (which is the column to compression codec map) in tableConfig in old style and asks for compression during initial segment creation
- Now sometime later in order to change compression, adds the same thing to `FieldConfig` but with different codec that they want to convert to but does not change the codec in the 2nd place in `noDictionaryConfig`
- We honor the `FieldConfig` and do the conversion using this feature. 
- User later backfills the segment (thus going through the generation path which actually uses the old style `noDictionaryConfig`) and ends up getting the segment in previous compression codec.

However, I checked that this will not happen. When `SegmentGeneratorConfig` is created, we actually unify / merge the configs (for the reason that migration hasn't been done) and so it will honor `FieldConfig`. Check the `extractCompressionCodecConfigsFromTableConfig` method which is invoked after the following code and thus given the preference

```
if (noDictionaryColumns != null) {
        this.setRawIndexCreationColumns(noDictionaryColumns);

        if (noDictionaryColumnMap != null) {
          Map<String, ChunkCompressionType> serializedNoDictionaryColumnMap = noDictionaryColumnMap.entrySet().stream()
              .collect(Collectors.toMap(Map.Entry::getKey, e -> ChunkCompressionType.valueOf(e.getValue())));
          this.setRawIndexCompressionType(serializedNoDictionaryColumnMap);
        }
      }
```

Reply:
Thanks for the clarification!

Reply:
FYI @somandal